### ISSUE TITLE
Accept more than one JWT secret for authentication

### ISSUE THREAD

**Issue by eric-brechemier** [PostgREST/postgrest#505]
In order to allow a smooth transition from one JWT secret to the next, without invalidating all existing authentication tokens at once, more than one JWT secret should be tried for authentication, typically:
- the JWT secret currently used for the signature of tokens
- the JWT secret used previously for the signature of tokens

In other words, the configuration should allow at least two JWT secrets, one for signature (and authentication), and one or more just for authentication.

**Comment by begriffs** [comment]
Makes sense, I can add that.

**Comment by begriffs** [comment]
Actually looking at this again is the complexity worth it? What if you just invalidated existing tokens? Then clients would request a new token and continue.

**Comment by eric-brechemier** [comment]
> Actually looking at this again is the complexity worth it?

Yes. There is no other way when using JWT for authentication (rather than password authentication over JWT).

Each JWT corresponds to a user role. Before we grant a new role to a user, with a new JWT, we need the current JWT to authenticate the user.

> What if you just invalidated existing tokens? 

Then existing users could only access the public API using the anonymous role.

> Then clients would request a new token and continue.

When doing so, they would simply become new users and would lose access to existing data.

**Comment by ruslantalpa** [comment]
@eric-brechemier especially your last comment about "become new users" does not make sense, why calling `login` a second time would make the user lose his data?

as a general note, i don't see why you would need to transition between two secrets.
You only change the jwt secret when you suspect it has been compromised in which case all the existing tokens are invalidated and each client needs to login again.

**Comment by eric-brechemier** [comment]
@ruslantalpa This is not merely `login`, as you assume. Generating a new secret is the equivalent of `signup` when one JWT = one identity.

> as a general note, i don't see why you would need to transition between two secrets.
You only change the jwt secret when you suspect it has been compromised (...)

We need to transition to a new secret in order to upgrade to newer versions of PostgREST at some point in the future, because the version that we are currently using has corrupted our secret (cf #495).

**Comment by ruslantalpa** [comment]
I think you are using the jwt in a strange way.
Let's assume there is this feature you are asking for. Your jwt secret get's compromised and you need to change it?
What do you do, since you said you need to have the old and the new one working at the same time because switching directly to a new one will make the old clients lose their data, at the same time, if you leave the old one in place, it means the attacker can create fake (but valid) tokens as long as he likes.

I don't know how your system is setup but i think there is a problem with it if changing a secret (private key) generates so many problems/inconveniences.  The only problem this should create is that the clients can not use their current token and ask for a new one.

**Comment by eric-brechemier** [comment]
@ruslantalpa I understand your incomprehension. Let's leave our particular use case apart.

Let's say that you create a policy to change your secret once a year. What happens when you update PostgREST's configuration? At soon as the new secret is active, all requests using JWT tokens emitted previously will fail.

What I suggest instead, is that, for the validity of previous tokens, the previous secret be tried as well to continue accepting valid tokens until they expire.

**Comment by ruslantalpa** [comment]
I think this situation needs to be solved in the client and not the backend.
In your scenario above you mentioned that the tokens expire, that means that sometimes in the future (no change on the server) the token for a particular client will expire and he will need to request a new token. That also means that on each request, the client needs to be able to handle "expired/invalid token" error responses. So why not just change the secret and have the clients get new tokens, just as if their token expired.

**Comment by eric-brechemier** [comment]
> So why not just change the secret and have the clients get new tokens, just as if their token expired.

Because that would cause all existing tokens to expire at the same time, resulting in a spike of activity with extra round trips for all connected users:

1. failed request
2. request for new token (which may require user intervention for authentication)
3. repeated request

**Comment by ruslantalpa** [comment]
I fail to see how that is a problem,
1. it would happen once a year
2. in regard to the roundtrip, from the client point of view, it's just an expired token, a usual thing that happens once in a while (far more often then once a year)
3. the extra login requests would not come in all at once as you said, they would come on the next get request of your clients, for them to come all at once, it means the all of your clients are constantly requesting data (more then once a second)
4. even if a lot of login requests come at once, a login procedure is much less expensive then a data fetch (no big tables involved) and i see no problem in handling such a spike when i got at leat hundreds of requests per second out of a micro instance in ec2, i.e. if you have 10k clients, your spike would be handled in 20s by the lowest hardware possible.

**Comment by eric-brechemier** [comment]
@ruslantalpa Here is another example: we have machines connected 24/7 to our API, unattended, with queries every 30s. Having every machine losing its authentication rights at the same time would result in support calls that we may easily handle individually, but not all at once.

**Comment by ruslantalpa** [comment]
Why would this result in you having to handle a support call? Why wouldn't the code on the client machine treat this case just like an expired token and request a new one?
Are you using long exp time for tokens and manually generating them and sending them to the client each time they expire?

**Comment by jackfirth** [comment]
Just another two cents here, but accepting multiple secrets is also important for rolling updates. If this "once a year secret rotation" were required across dozens or hundreds of postgrest instances that were gradually redeployed (as opposed to all at once) behind a load balancer, then it would be perfectly possible for requests to an old instance of postgrest to end up hitting an old instance one request and a new one on the next. This could happen in reverse too: a client authenticates against a new instance with the new secret then subsequent requests get load-balanced to old instances.

A set of secrets with one designated as "preferred" would solve this problem by turning it into a two-phase rollout. In phase one, all instances are upgraded to accept the new secret but still issue tokens with the old secret. Then in phase two, all instances are upgraded to prefer the new secret instead of the old one. This way staged rollouts can be done without interrupting client auth.

**Comment by eric-brechemier** [comment]
> Are you using long exp time for tokens and manually generating them and sending them to the client each time they expire?

Nearly: we are validating authorized users over the phone.

> Why wouldn't the code on the client machine treat this case just like an expired token and request a new one?

The client machine runs unattended. Human intervention is required to authenticate and obtain a new token.

**Comment by eric-brechemier** [comment]
> A set of secrets with one designated as "preferred" would solve this problem by turning it into a two-phase rollout.

Exactly. It feels like this would be easier to configure at PostgREST level, but I thought of a different way to put it in place by spawning two separate instances of PostgREST (each with its own secret) and falling back to the one with lower priority when JWT authentication fails on the one with higher priority.

**Comment by ruslantalpa** [comment]
ok, so your client calls you, you generate a jwt, give it to them and they save it as a file and the client code uses it. 
Why not instead give them a normal user/password that they save in the same file, and when the client box boots up, in the beginning, it does an additional request to get a jwt an then use it instead of expecting to be handed a preexisting jwt.

**Comment by eric-brechemier** [comment]
> ok, so your client calls you, you generate a jwt, give it to them and they save it as a file and the client code uses it. 

No. For the record: this particular client app is running in a web page which calls our API to get a JWT and saves it in offline storage of the browser. The page displays the role encoded in the JWT. We talk with a representative of the organization over the phone and grant the role privileges for this organization. This procedure could be automated but we have limited resources and lots of features more valuable for the product in our development pipe.

> Why not instead give them a normal user/password (...)

I have strong feelings against passwords. We are getting really off topic though.

@begriffs @ruslantalpa What do you think of rolling secrets at the end of this discussion? Is it in scope for a future version of PostgREST or should we plan to implement it at the level of a proxying Web server such as nginx instead?

**Comment by begriffs** [comment]
This feature makes sense when getting a new token is difficult (as it is for Eric's clients), or during a rolling deployment as Jack points out. In the former case you might argue that there is a better process for handling tokens, but the latter case seems especially difficult to accomplish without allowing a secondary secret.

Doing a rolling deploy with only one secret could cause severe authentication churn in the clients. Imagine that a request comes in and routes to a new-secret server. The server rejects it and tells the client to authenticate. The client makes an authentication request that happens to route to an old-secret server, acquiring another old token! Alternately the request gets a new-secret token but subsequent requests may still route to an old-secret server leading to the same problem all over again.

With a secondary secret you can update servers using this process:

1. Through rolling deployment: Add the new secret as secondary, leave the old primary as primary.
1. Once all servers have updated, change the JWT issuing code to sign any new tokens with the new (secondary) secret. Depending on JWT expiration times in the outstanding tokens it may be a while before clients get new tokens.
1. Through rolling deployment: When you feel enough clients are on the new secret, promote the secondary secret to primary.

In short I'm starting to favor allowing secondary secrets. Thoughts?

**Comment by ruslantalpa** [comment]
a) the rolling update case - any load balancer has a thing called "sticky session", so your scenario will never happen. This is overengeneering a solution for o.1% deployments of postgrest (i'd like to see someone overload there postgrest box) for a situation that will happen once every 6 months and would last for less then 1s.

b) getting a token is difficult case - i really don't see why this is difficult and why it's implemented in this manner, why a simple password (doesn't mean short) is not enough  any why it's any less secure then a jwt token. It seems to me as there is no real technical reason/need for this feature it's just a case of "we have other features to implement until we get to this bit". If there is ever such an edge case that there is no way around this 2 simultaneous secrets (i haven't been convinced one exists yet), the proxy could decode the old jwt and reencode with the new secret on the fly. I would also question if there is a real need for this secret change once every six months in typical deployments. Are you really being targeted by such skilled hackers that they were able to get through all the firewalls and os securities and got the the file on your hdd containing the jwt secret or do you have so little trust in the employees and think that they would still this key and if so, why do the have access to it in the first place.

**Comment by jackfirth** [comment]
Sticky sessions are not guarantees. Machines that sessions are connected to can fail, connections can be closed and reopened with the same JWTs, and keeping state in load balancers hurts horizontal scaling. Keeping deployments of services stateless and "identity-less" (not having to care what instance a request hits) is very important for managing deployment complexity. Not to mention many users would likely rather not be required to configure sticky sessions at all - it's unnecessary complexity for this problem.

Rolling deployments are *not* something that occur in a matter of seconds. In a large scale deployment, upgrades are likely to be spread out across several days with heavy testing at various stages of the rollout (i.e., rerunning end-to-end and performance tests once 5% of services are upgraded, then again at 20%, then 50%, etc.) . This ensures bad configuration changes don't take the entire service down. Cycling the secret used by Postgrest is also not something that will only ever happen every six months, in many security-conscious systems secret keys are changed on a weekly basis, not annually.

This is the kind of feature large-scale use of Postgrest would care about a lot, but medium and small scale users will likely never care about or even consider in depth.

**Comment by ruslantalpa** [comment]
@jackfirth your last paragraph i the important one, the first two apply only to G/FB and organisations like that and they will never use postgrest (or are capable of rewriting it however they want).  So i guess i should have said o.000001% of deployments....

**Comment by spinus** [comment]
Hey guys. I was just advised today that JWT tokens kept in database settings is not that cool as any user could read that and probably would be better to keep JWT tokens in a table which can be properly secured. 

This kind of relates to this issue, we could implement feature to postgrest to use tokens from specific table and than multuple tokens would be supported and overal setup would be both more secure (not exposing tokens to all db users) and flexible (as people mention here, tokens could be added and removed easily). Let me know what do you think.

**Comment by ppKrauss** [comment]
Checking concept: is the  second token "just for authentication" of the introduction, a [*refresh token*](https://auth0.com/docs/tokens/refresh-token)? ([other example](https://github.com/svlada/springboot-security-jwt/blob/master/src/main/java/com/svlada/security/endpoint/RefreshTokenEndpoint.java)).

**Comment by eric-brechemier** [comment]
@ppKrauss Yes. The old token would be used as a refresh token for the introduction of the new token.

**Comment by ppKrauss** [comment]
The most popular libraries of [jwt.io](https://jwt.io/#libraries-io) offer the *refresh token feature*: it is a kind of [design pattern](https://en.wikipedia.org/wiki/Software_design_pattern) in JWT community... So, I think that this issue (the @eric-brechemier suggestion) can use this concept to simplify discussion, and perhaps adopt a "standard way" to do it. 

Summarizing, the design of endpoints in "JWT with refresh token feature" pattern is:

* The *login endpoint* outputs two tokens, the *usual JWT token* (that will expires) and the *refresh-token*.
* A *refresh-token endpoint* can be used to output a new *usual token* (with new expiration date) when authenticated by the input *refresh-token*.

**Comment by eric-brechemier** [comment]
@ppKrauss the question is: how to you renew the refresh-token (the one that does not expire), if you really have to. I described a possible approach [here](https://github.com/begriffs/postgrest/issues/809#issuecomment-281041793):

* use the latest secret to sign new never-expiring JWT tokens
* allow never-expiring JWT tokens signed with older secrets to be used for some time to request a new never-expiring JWT token

**Comment by ppKrauss** [comment]
**General suggestion**: let's use the standard jargon and concepts of [RFC 6749 section-1.5](https://tools.ietf.org/html/rfc6749#section-1.5): the name is *Refresh Token*... (we can abbreviate to "token" the name of *Access Token*).

... as [I explained above](#issuecomment-281066281) the token and refresh-token are supplied by login... For use it, needs specific "refresh-token *endpoint*", where occurs specific "refresh-token *event*" (if authenticates refresh-token it issues a new access-token). 

-----

PS: @eric-brechemier  the name "never-expiring token" make confusion, "eternity" is not an obligation, the only constraint of *Refresh Token* is [`rfshtoken_valid_until  > token_valid_until`](https://github.com/begriffs/postgrest/issues/809#issuecomment-281118234). Coping the quotes of [this SO popular answer](http://stackoverflow.com/a/7209263/287948):

> Refresh tokens... mitigates the risk of a long-lived access_token leaking (query param in a log file on an insecure resource server, beta or poorly coded resource server app, JS SDK client on a non https site that puts the access_token in a cookie, etc)

**Comment by spinus** [comment]
I know it's a little bit a side topic but could we use name "refresh_token" for Refresh Token and "access_token" for access token? I think it will help people to follow implementation and understand the code. 

I had a lot of journeys when I started using some software or reading the code and they used different names (sometimes the convention wasn't even consistent inside the app) and that was confusing and frustrating when I found that "token" is actually "secret" and "secret" is "private_bit" or something like that.

Also, IMHO, "rfsh_token" takes me longer to write then "refresh_token", I guess also harder to remember :-)
What do you think?

**Comment by eric-brechemier** [comment]
@ppKrauss I stand corrected :) Thank you for the [reference](https://tools.ietf.org/html/rfc6749#section-1.5); I agree that the terminology is useful.

PostgREST (v0.3 at least) plays two different roles:

* Resource Server, which accepts "Access Token" JWT 
* Authorization Server, which emits "Refresh Token" JWT

Likewise, the "Access Token" and "Refresh Token" are actually one and the same in current architecture.

My suggestion, rephrased, is for PostgREST to support a list of secrets instead of a single secret, and to consider that JWTs signed with the most recent secrets are allowed both as "Access Token" and "Refresh Token" (current behavior)  while JWTs signed with older secrets are only allowed as "Refresh Token".

**Comment by begriffs** [comment]
From this thread it seems like we ought to support both access- and refresh-tokens. Anyone want to volunteer to implement this?

**Comment by eric-brechemier** [comment]
@begriffs can you lay down some guide lines about how you think this should be implemented? I am wondering in particular how much should be done in SQL alone, and how much in Haskell.

**Comment by PierreRochard** [comment]
@eric-brechemier if the complexity is comparable for SQL vs Haskell implementation, as a humble hobby user my vote is for SQL. 😃

**Comment by begriffs** [comment]
OK so what if postgrest itself can simply accept two tokens in the config file, and if one fails when validating a client request it tries the other? Then people can make the various endpoints in SQL for producing refresh/access tokens. Would that suffice?

**Comment by ppKrauss** [comment]
Hi   @begriffs, there are a good clue in the [pgjwt](https://github.com/michelp/pgjwt) extension, that was  cited by you at [docs/auth.html#jwt-generation](https://postgrest.com/en/v0.4/auth.html#jwt-generation) (see [test.sql](https://github.com/michelp/pgjwt/blob/master/test.sql)).

**Comment by eric-brechemier** [comment]
> Would that suffice?

yes!

**Comment by ppKrauss** [comment]
Hi @begriffs, are there an "official data-model" about it?  UML class-diagram and SQL table definitions?

PS: I like to use http://yuml.me/ and can help with documentation.

**Comment by begriffs** [comment]
By the way, I've been working on JWT changes on a local branch. The changes will make it easier to use multiple keys.

**Comment by eric-brechemier** [comment]
@begriffs I am working on additional unit tests for JWT secrets, in the context of #876 and #877, currently on top of `master`. I will synchronize with your branch instead if you push it remotely.

**Comment by begriffs** [comment]
More tests, that's great!

My work on this feature is being sponsored by a company, so I will make it public after their review and payment.

**Comment by begriffs** [comment]
Just an update here, the company accepted the work based on a demo and I'm awaiting payment now.

**Comment by begriffs** [comment]
^^^ The switch to another JWT library is now available in the linked PR.

When it is suitable to merge we can start working on allowing multiple keys in the config file.

**Comment by begriffs** [comment]
The JWT branch is merged now!

In case anybody wants to help out with this multiple-key issue, what you need to do is have it parse a [JWKSet](https://hackage.haskell.org/package/jose-0.6.0.3/docs/Crypto-JOSE-JWK.html#g:5) rather than simply JWK in [parseJWK](https://github.com/begriffs/postgrest/blob/master/src/PostgREST/Auth.hs#L77). It should have a built in instance to parse it from JSON, so changing the type signature of the parseJWK function should do the trick. However you then have to follow the compiler errors to make the validation code work with a JWKSet.

**Comment by elimisteve** [comment]
Anyone else interested in a feature that uses a different JWT signing key for different groups of users? I'd like a stolen signing key to only allow the attacker to delete some rows, not all rows :-).

Though if all the keys are right next to each other then using a bunch of different signing keys doesn't help.

Any thoughts on making this more resilient to attack under the compromised JWT signing key scenario (other than "monitor servers for attacks + change the signing key ASAP")?

**Comment by steve-chavez** [comment]
@elimisteve If your tables are protected with RLS, an attacker would only be able to delete his own rows(the ones that belong to the stolen user account key) and not all rows.

**Comment by elimisteve** [comment]
@steve-chavez If the signing key is stolen then the attacker could create new JWTs with for each (especially high-privilege) user, then use those permissions granted by the self-minted JWTs to essentially do whatever all users combined could do, though perhaps requiring multiple JWTs to do it.
